content/blog/2019-10-23_react-typescript-starter.md

```diff
@@ -0,0 +1,464 @@
+        app: "./src/index.tsx"
+    },
+    output: {
+        path: path.resolve(__dirname, "dist"),
+        filename: "app.[hash].js"
+    },
+    devtool: "eval-source-map",
+    resolve: {
+        extensions: [".js", ".jsx", ".ts", ".tsx"]
+    },
+    module: {
+        rules: []
+    },
+};
+```
+This means we are using `src/index.tsx` as the entry point (e.g. what Webpack is loading) and the output will be created in
+the `dist` directory, containing the hash in the filename for caching reasons.
+We use `eval-source-map` as `devtool` value as this is the best setting for development, You can see all the different values
+in the [Webpack devtool documentation](https://webpack.js.org/configuration/devtool/#devtool). Finally, we let Webpack know that
+it should try to load imports with no extensions as `.js`, `.jsx`, .`ts` and `.tsx` (so we can do `import something from "../index"`) and add a script to build our project.
+
+```diff
+diff --git a/package.json b/package.json
+index eb50375..1bc2df7 100644
+--- a/package.json
++++ b/package.json
+@@ -1,4 +1,8 @@
+ {
++  "scripts": {
++    "build": "webpack",
++  },
+   "devDependencies": {
+     "typescript": "^3.6.4",
+     "webpack": "^4.41.2",
+```
+
+Running that script (`yarn build`) will create the `dist` folder containing a file with a name like `app.5730f5cdf8cb70458282.js`.
+Since we didn't write any code yet, it currently only contains Webpack own code.
+
+The commit for those changes is in <https://github.com/Keats/webpack-react-typescript/commit/48fc3f342c3f4f53c30af1b457d90351f442b4f0>.
+
+### Loaders and plugin
+
+In this article we are aiming for a simple SPA without SSR so let's setup a super basic React application in order to actually bundle something.
+
+```bash
+$ yarn add react react-dom
+$ yarn add @types/react @types/react-dom --dev
+```
+
+React doesn't ship the TypeScript type definitions by default so we need to download them manually.
+
+The simplest React code we can write in `src/index.tsx` is:
+
+```tsx
+import React from "react";
+import ReactDOM from "react-dom";
+
+ReactDOM.render(<h1>Hello world</h1>, document.getElementById("root"));
+```
+
+This will render the React app in the HTML element with id `root`, which, as some of you might have noticed, doesn't exist yet.
+
+There is a plugin that will handle the HTML for us: [html-webpack-plugin](https://github.com/jantimon/html-webpack-plugin).
+
+```bash
+$ yarn add html-webpack-plugin --dev
+```
+
+Enabling it in our `webpack.config.js` is trivial:
+
+```diff
+diff --git a/webpack.config.js b/webpack.config.js
+index 2267dfd..bcbe494 100644
+--- a/webpack.config.js
++++ b/webpack.config.js
+@@ -1,4 +1,5 @@
+ const path = require("path");
++const HtmlWebpackPlugin = require("html-webpack-plugin");
+ 
+ module.exports = {
+     context: __dirname,
+@@ -17,4 +18,7 @@ module.exports = {
+     module: {
+         rules: []
+     },
++    plugins: [
++        new HtmlWebpackPlugin({ template: "index.html" }),
```

Comment:
I never used this plugin. The index.html is always dynamically generated to add the meta you want, so you cannot treat it as a static file.

Reply:
Je comprends pas trop ce que tu veux dire par là. C'est un template donc tu mets ce que tu veux. Quel `meta` tu changes de manière dynamique?

Reply:
meta description & co pour les SEO. Faut que ce soit généré par le server et dispo sans js dans le premier html.

Reply:
Ah je vois, c'est pas trop mon cas la vu que osef de SSR/SEO pour cet article. Tu fais ca comment toi?

Reply:
rien de spécial. J'utilise juste pas ce plugin xD dans mon index.html, je link vers le fichier js, et si c'est un nombre de chunks avec des hashes non prédictables, je fait un wrapper au dessus de webpack pour upload les fichiers dans GCS et changer les env vars contenant les noms de fichier. 
osef, ignore le comment :)